<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>FASTA formatted files · FASTX.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../../assets/documenter.js"></script><script src="../../../siteinfo.js"></script><script src="../../../../versions.js"></script><link href="../../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>FASTX.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../../">Home</a></li><li><span class="toctext">Manual</span><ul><li class="current"><a class="toctext" href>FASTA formatted files</a><ul class="internal"><li><a class="toctext" href="#Readers-and-Writers-1">Readers and Writers</a></li></ul></li><li><a class="toctext" href="../fastq/">FASTQ formatted files</a></li></ul></li><li><span class="toctext">API Reference</span><ul><li><a class="toctext" href="../../../reference/fasta/">FASTA formatted files</a></li><li><a class="toctext" href="../../../reference/fastq/">FASTQ formatted files</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>Manual</li><li><a href>FASTA formatted files</a></li></ul><a class="edit-page" href="https://github.com/BioJulia/FASTX.jl/blob/master/docs/src/manual/fasta.jl"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>FASTA formatted files</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="IO-FASTA-formatted-files-1" href="#IO-FASTA-formatted-files-1">IO - FASTA formatted files</a></h1><p>FASTA is a text-based file format for representing biological sequences. A FASTA file stores a list of sequence records with name, description, and sequence.</p><p>The template of a sequence record is:</p><pre><code class="language-none">&gt;{name} {description}?
{sequence}</code></pre><p>Here is an example of a chromosomal sequence:</p><pre><code class="language-none">&gt;chrI chromosome 1
CCACACCACACCCACACACCCACACACCACACCACACACCACACCACACC
CACACACACACATCCTAACACTACCCTAACACAGCCCTAATCTAACCCTG</code></pre><h2><a class="nav-anchor" id="Readers-and-Writers-1" href="#Readers-and-Writers-1">Readers and Writers</a></h2><p>The reader and writer for FASTA formatted files, are found within the <code>BioSequences.FASTA</code> submodule.</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="FASTX.FASTA.Reader" href="#FASTX.FASTA.Reader"><code>FASTX.FASTA.Reader</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">FASTA.Reader(input::IO; index = nothing)</code></pre><p>Create a data reader of the FASTA file format.</p><p><strong>Arguments</strong></p><ul><li><code>input</code>: data source</li><li><code>index=nothing</code>: filepath to a random access index (currently <em>fai</em> is supported)</li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/aed3a552c400b95b7e46f709de8130cf25607338/src/fasta/reader.jl#L13-L21">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="FASTX.FASTA.Writer" href="#FASTX.FASTA.Writer"><code>FASTX.FASTA.Writer</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">FASTA.Writer(output::IO; width=70)</code></pre><p>Create a data writer of the FASTA file format.</p><p><strong>Arguments</strong></p><ul><li><code>output</code>: data sink</li><li><code>width=70</code>: wrapping width of sequence characters</li></ul></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/aed3a552c400b95b7e46f709de8130cf25607338/src/fasta/writer.jl#L4-L12">source</a></section><p>They can be created with IOStreams:</p><pre><code class="language-">using FASTX

r = FASTA.Reader(open(&quot;MyInput.fasta&quot;, &quot;r&quot;))
w = FASTA.Writer(open(&quot;MyFile.fasta&quot;, &quot;w&quot;))</code></pre><p>Alternatively, <code>Base.open</code> is overloaded with a method for conveinience:</p><pre><code class="language-">r = open(FASTA.Reader, &quot;MyInput.fasta&quot;)
w = open(FASTA.Writer, &quot;MyFile.fasta&quot;)</code></pre><p>Usually sequence records will be read sequentially from a file by iteration.</p><pre><code class="language-">reader = open(FASTA.Reader, &quot;hg38.fa&quot;)
for record in reader
    # Do something
end
close(reader)</code></pre><p>But if the FASTA file has an auxiliary index file formatted in fai, the reader supports random access to FASTA records, which would be useful when accessing specific parts of a huge genome sequence:</p><pre><code class="language-">reader = open(FASTA.Reader, &quot;sacCer.fa&quot;, index=&quot;sacCer.fa.fai&quot;)
chrIV = reader[&quot;chrIV&quot;]  # directly read sequences called chrIV.
close(reader)</code></pre><p>Reading in a sequence from a FASTA formatted file will give you a variable of type <code>FASTA.Record</code>.</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="FASTX.FASTA.Record" href="#FASTX.FASTA.Record"><code>FASTX.FASTA.Record</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">FASTA.Record()</code></pre><p>Create an unfilled FASTA record.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/aed3a552c400b95b7e46f709de8130cf25607338/src/fasta/record.jl#L14-L18">source</a><div><div><pre><code class="language-none">FASTA.Record(data::Vector{UInt8})</code></pre><p>Create a FASTA record object from <code>data</code>.</p><p>This function verifies and indexes fields for accessors. Note that the ownership of <code>data</code> is transferred to a new record object.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/aed3a552c400b95b7e46f709de8130cf25607338/src/fasta/record.jl#L23-L30">source</a><div><div><pre><code class="language-none">FASTA.Record(str::AbstractString)</code></pre><p>Create a FASTA record object from <code>str</code>.</p><p>This function verifies and indexes fields for accessors.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/aed3a552c400b95b7e46f709de8130cf25607338/src/fasta/record.jl#L41-L47">source</a><div><div><pre><code class="language-none">FASTA.Record(identifier, sequence)</code></pre><p>Create a FASTA record object from <code>identifier</code> and <code>sequence</code>.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/aed3a552c400b95b7e46f709de8130cf25607338/src/fasta/record.jl#L60-L64">source</a><div><div><pre><code class="language-none">FASTA.Record(identifier, description, sequence)</code></pre><p>Create a FASTA record object from <code>identifier</code>, <code>description</code> and <code>sequence</code>.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/aed3a552c400b95b7e46f709de8130cf25607338/src/fasta/record.jl#L69-L73">source</a></section><p>Various getters and setters are available for <code>FASTA.Record</code>s:</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="FASTX.FASTA.hasidentifier" href="#FASTX.FASTA.hasidentifier"><code>FASTX.FASTA.hasidentifier</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">hasidentifier(record::Record)</code></pre><p>Checks whether or not the <code>record</code> has an identifier.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/aed3a552c400b95b7e46f709de8130cf25607338/src/fasta/record.jl#L166-L170">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="FASTX.FASTA.identifier" href="#FASTX.FASTA.identifier"><code>FASTX.FASTA.identifier</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">identifier(record::Record)::String</code></pre><p>Get the sequence identifier of <code>record</code>.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/aed3a552c400b95b7e46f709de8130cf25607338/src/fasta/record.jl#L153-L157">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="FASTX.FASTA.hasdescription" href="#FASTX.FASTA.hasdescription"><code>FASTX.FASTA.hasdescription</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">hasdescription(record::Record)</code></pre><p>Checks whether or not the <code>record</code> has a description.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/aed3a552c400b95b7e46f709de8130cf25607338/src/fasta/record.jl#L196-L200">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="FASTX.FASTA.description" href="#FASTX.FASTA.description"><code>FASTX.FASTA.description</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">description(record::Record)::String</code></pre><p>Get the description of <code>record</code>.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/aed3a552c400b95b7e46f709de8130cf25607338/src/fasta/record.jl#L183-L187">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="FASTX.FASTA.hassequence" href="#FASTX.FASTA.hassequence"><code>FASTX.FASTA.hassequence</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">hassequence(record::Record)</code></pre><p>Checks whether or not a sequence record contains a sequence.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/aed3a552c400b95b7e46f709de8130cf25607338/src/fasta/record.jl#L245-L249">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="FASTX.FASTA.sequence-Tuple{FASTX.FASTA.Record,UnitRange{Int64}}" href="#FASTX.FASTA.sequence-Tuple{FASTX.FASTA.Record,UnitRange{Int64}}"><code>FASTX.FASTA.sequence</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-none">sequence(record::Record, [part::UnitRange{Int}])</code></pre><p>Get the sequence of <code>record</code>.</p><p>This function infers the sequence type from the data. When it is wrong or unreliable, use <code>sequence(::Type{S}, record::Record)</code>.  If <code>part</code> argument is given, it returns the specified part of the sequence.</p></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/aed3a552c400b95b7e46f709de8130cf25607338/src/fasta/record.jl#L230-L238">source</a></section><p>To write a <code>BioSequence</code> to FASTA file, you first have to create a <code>FASTA.Record</code>:</p><div><pre><code class="language-julia">using BioSequences
x = dna&quot;aaaaatttttcccccggggg&quot;
rec = FASTA.Record(&quot;MySeq&quot;, x)
w = open(FASTA.Writer, &quot;MyFile.fasta&quot;)
write(w, rec)
close(w)</code></pre><pre><code class="language-none">WARNING: using BioSequences.FASTA in module ex-fasta conflicts with an existing identifier.</code></pre></div><p>As always with julia IO types, remember to close your file readers and writer after you are finished.</p><p>Using <code>open</code> with a do-block can help ensure you close a stream after you are finished.</p><pre><code class="language-">open(FASTA.Reader, &quot;hg38.fa&quot;) do reader
    for record in reader
        # Do something
    end
end</code></pre><p><em>This page was generated using <a href="https://github.com/fredrikekre/Literate.jl">Literate.jl</a>.</em></p><footer><hr/><a class="previous" href="../../../"><span class="direction">Previous</span><span class="title">Home</span></a><a class="next" href="../fastq/"><span class="direction">Next</span><span class="title">FASTQ formatted files</span></a></footer></article></body></html>
