<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>FASTQ formatted files · FASTX.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../index.html"><img class="logo" src="../../assets/logo.png" alt="FASTX.jl logo"/></a><h1>FASTX.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><span class="toctext">Manual</span><ul><li><a class="toctext" href="../fasta/">FASTA formatted files</a></li><li class="current"><a class="toctext" href>FASTQ formatted files</a><ul class="internal"><li><a class="toctext" href="#Readers-and-Writers-1">Readers and Writers</a></li><li><a class="toctext" href="#Quality-encodings-1">Quality encodings</a></li></ul></li></ul></li><li><span class="toctext">Library</span><ul><li><a class="toctext" href="../../lib/public/">Public</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>Manual</li><li><a href>FASTQ formatted files</a></li></ul><a class="edit-page" href="https://github.com/BioJulia/FASTX.jl/blob/master/docs/src/manual/fastq.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>FASTQ formatted files</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="FASTQ-formatted-files-1" href="#FASTQ-formatted-files-1">FASTQ formatted files</a></h1><p>FASTQ is a text-based file format for representing DNA sequences along with qualities for each base. A FASTQ file stores a list of sequence records in the following format:</p><pre><code class="language-none">@{name} {description}?
{sequence}
+
{qualities}</code></pre><p>Here is an example of one record from a FASTQ file:</p><pre><code class="language-none">@FSRRS4401BE7HA
tcagTTAAGATGGGAT
+
###EEEEEEEEE##E#</code></pre><h2><a class="nav-anchor" id="Readers-and-Writers-1" href="#Readers-and-Writers-1">Readers and Writers</a></h2><p>The reader and writer for FASTQ formatted files, are found within the FASTQ module of FASTX.</p><ul><li><a href="../../lib/public/#FASTX.FASTQ.Reader"><code>FASTQ.Reader</code></a></li><li><a href="../../lib/public/#FASTX.FASTQ.Writer"><code>FASTQ.Writer</code></a></li></ul><p>They can be created with IOStreams:</p><pre><code class="language-jlcon">using FASTX

r = FASTQ.Reader(open(&quot;my-reads.fastq&quot;, &quot;r&quot;))
w = FASTQ.Writer(open(&quot;my-output.fastq&quot;, &quot;w&quot;))</code></pre><p>Alternatively, <code>Base.open</code> is overloaded with a method for conveinience:</p><pre><code class="language-jlcon">r = open(FASTQ.Reader, &quot;my-reads.fastq&quot;)
w = open(FASTQ.Writer, &quot;my-out.fastq&quot;)</code></pre><p>Note that <a href="../../lib/public/#FASTX.FASTQ.Reader"><code>FASTQ.Reader</code></a> does not support line-wraps within sequence and quality. Usually sequence records will be read sequentially from a file by iteration.</p><pre><code class="language-jlcon">reader = open(FASTQ.Reader, &quot;my-reads.fastq&quot;)
for record in reader
    ## Do something
end
close(reader)</code></pre><p>You can also overwrite records in a while loop to avoid excessive memory allocation.</p><pre><code class="language-jlcon">reader = open(FASTQ.Reader, &quot;my-reads.fastq&quot;)
record = FASTQ.Record()
while !eof(reader)
    read!(reader, record)
    ## Do something.
end</code></pre><p>Reading in a record from a FASTQ formatted file will give you a variable of type <a href="../../lib/public/#FASTX.FASTQ.Record"><code>FASTQ.Record</code></a>.</p><p>Various getters and setters are available for <a href="../../lib/public/#FASTX.FASTQ.Record"><code>FASTQ.Record</code></a>s:</p><ul><li><a href="../../lib/public/#FASTX.FASTQ.hasidentifier"><code>FASTQ.hasidentifier</code></a></li><li><a href="../../lib/public/#FASTX.identifier"><code>FASTQ.identifier</code></a></li><li><a href="../../lib/public/#FASTX.FASTQ.hasdescription"><code>FASTQ.hasdescription</code></a></li><li><a href="../../lib/public/#FASTX.description"><code>FASTQ.description</code></a></li><li><a href="../../lib/public/#FASTX.FASTQ.hassequence"><code>FASTQ.hassequence</code></a></li><li><a href="../../lib/public/#FASTX.sequence"><code>FASTQ.sequence</code></a></li><li><a href="../../lib/public/#FASTX.FASTQ.hasquality"><code>FASTQ.hasquality</code></a></li><li><a href="../../lib/public/#FASTX.FASTQ.quality"><code>FASTQ.quality</code></a></li></ul><p>To write a <code>BioSequence</code> to FASTQ file, you first have to create a <a href="../../lib/public/#FASTX.FASTQ.Record"><code>FASTQ.Record</code></a>:</p><p>As always with julia IO types, remember to close your file readers and writer after you are finished.</p><p>Using <code>open</code> with a do-block can help ensure you close a stream after you are finished.</p><pre><code class="language-jlcon">open(FASTQ.Reader, &quot;my-reads.fastq&quot;) do reader
    for record in reader
        ## Do something
    end
end</code></pre><h2><a class="nav-anchor" id="Quality-encodings-1" href="#Quality-encodings-1">Quality encodings</a></h2><p>FASTQ records have a quality string which have platform dependent encodings. The FASTQ submodule has encoding and decoding support for the following quality encodings. These can be used with a <a href="../../lib/public/#FASTX.FASTQ.quality"><code>FASTQ.quality</code></a> method, to ensure the correct quality score values are extracted from your FASTQ quality strings. </p><ul><li><a href="../../lib/public/#FASTX.FASTQ.SANGER_QUAL_ENCODING"><code>FASTQ.SANGER_QUAL_ENCODING</code></a></li><li><a href="../../lib/public/#FASTX.FASTQ.SOLEXA_QUAL_ENCODING"><code>FASTQ.SOLEXA_QUAL_ENCODING</code></a></li><li><a href="../../lib/public/#FASTX.FASTQ.ILLUMINA13_QUAL_ENCODING"><code>FASTQ.ILLUMINA13_QUAL_ENCODING</code></a></li><li><a href="../../lib/public/#FASTX.FASTQ.ILLUMINA15_QUAL_ENCODING"><code>FASTQ.ILLUMINA15_QUAL_ENCODING</code></a></li><li><a href="../../lib/public/#FASTX.FASTQ.ILLUMINA18_QUAL_ENCODING"><code>FASTQ.ILLUMINA18_QUAL_ENCODING</code></a></li></ul><footer><hr/><a class="previous" href="../fasta/"><span class="direction">Previous</span><span class="title">FASTA formatted files</span></a><a class="next" href="../../lib/public/"><span class="direction">Next</span><span class="title">Public</span></a></footer></article></body></html>
